### PR TITLE
uct/ugni: fix initialization

### PR DESCRIPTION
**Pull request by hjelmn**
We were getting user complaints in Open MPI because UCX was failing to
initialize when used on a Cray system. This commit fixes the issue but
changing the way the CDM is calculated to match what is used in Open
MPI. This eliminates the need for using Cray's PMI at all. I am taking
away a bit of uniqueness from the CDM space because this might have to
live side-by-side with btl/ugni.

Signed-off-by: Nathan Hjelm <hjelmn@lanl.gov>

## What
Fixes UCX initialization on Cray systems.

## Why ?
Support for Cray systems is broken when not using Cray PMI.

## How ?
Create the CDM identifier using something other than the PMI properties. Also fix the domain_id field in the CDM struct. uint16_t was wrong and was likely being overflowed.

### CODE REVIEW

**Comment by hjelmn** [comment]
We still strongly recommend against using UCX on Cray systems with Open MPI (we will probably never recommend it) but this fixes one glaring issue. I am tracking another bug and will likely just open an issue for that one as I don't have the time to fix ucp bugs.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8387/ for details (Mellanox internal link).

**Comment by hppritcha** [comment]
bot:mellanox:retest

**Comment by brminich** [comment]
jenkins env issues

bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8393/ for details (Mellanox internal link).

**Comment by hjelmn** [comment]
Updated to correct the domain_id type in one other struct. Runs cleanly with btl/uct now.

File: src/uct/ugni/base/ugni_device.c [hunk 1]
```diff
@@ -23,20 +25,13 @@
...
 } uct_ugni_job_info_t;
 
 static uct_ugni_job_info_t job_info = {
-    .ptag               = 0,
-    .cookie             = 0,
-    .pmi_num_of_ranks   = 0,
-    .pmi_rank_id        = 0,
     .num_devices        = -1,
-    .initialized        = 0,
 };
```
Comment (hjelmn) [inline src/uct/ugni/base/ugni_device.c]: Just as a note here. I removed any extraneous line setting a struct member to 0. C requires the compiler to initialize any uninitialized memory in the data sections to 0 (both public and private) so .ptag = 0 does nothing.

**Comment by hjelmn** [comment]
One more fix. @hppritcha They were not using FMA_SHARED and since they do not expose the CDM mode as a parameter they have to always set it. Without it the result is predicable:

```
[1544045407.678697] [nid00043:25120:0]    ugni_device.c:441  UCX  ERROR GNI_CdmAttach failed, Error status: GNI_RC_ERROR_RESOURCE
```

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8394/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
@hjelmn - my understanding there are some failures ? how do we want to proceed ?

**Comment by hjelmn** [comment]
@shamisp The failures are on systems without ugni from what I can see. I can't see why the internal Mellanox stuff is failing but this is working fine on our XC-40.

File: src/uct/ugni/base/ugni_device.c [hunk 2]
```diff
@@ -389,26 +372,58 @@ ucs_status_t uct_ugni_iface_get_dev_address(uct_iface_t *tl_iface, uct_device_ad
     return UCS_OK;
 }
 
+static int uct_ugni_next_power_of_two_inclusive (int value)
```
Comment (shamisp) [inline src/uct/ugni/base/ugni_device.c]: @yosefe should it go to ucs as an utility ?

**Comment by brminich** [comment]
failures should be fixed by #3083 

bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8414/ for details (Mellanox internal link).

**Comment by brminich** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8415/ for details (Mellanox internal link).

File: src/uct/ugni/base/ugni_device.c (hunk 2 above)
Comment (yosefe) [reply src/uct/ugni/base/ugni_device.c]: up to @hjelmn

**Pull request merged**